ouroboros-network/src/Data/Sequence/Strict.hs

```diff
@@ -0,0 +1,221 @@
+
+-- | 'scanl' is similar to 'foldl', but returns a sequence of reduced
+-- values from the left:
+--
+-- > scanl f z (fromList [x1, x2, ...]) = fromList [z, z `f` x1, (z `f` x1) `f` x2, ...]
+scanl :: (a -> b -> a) -> a -> StrictSeq b -> StrictSeq a
+scanl f !z0 (StrictSeq xs) = StrictSeq $ forceElemsToWHNF $ Seq.scanl f z0 xs
+-- scanl f z0 xs = z0 <| snd (mapAccumL (\ x z -> let x' = f x z in (x', x')) z0 xs)
+
+{-------------------------------------------------------------------------------
+  Sublists
+-------------------------------------------------------------------------------}
+
+-- | \( O(i) \) where \( i \) is the prefix length.  @'dropWhileL' p xs@ returns
+-- the suffix remaining after @'takeWhileL' p xs@.
+dropWhileL :: (a -> Bool) -> StrictSeq a -> StrictSeq a
+dropWhileL p (StrictSeq xs) = StrictSeq (Seq.dropWhileL p xs)
+
+-- | \( O(i) \) where \( i \) is the suffix length.  @'dropWhileR' p xs@ returns
+-- the prefix remaining after @'takeWhileR' p xs@.
+--
+-- @'dropWhileR' p xs@ is equivalent to @'reverse' ('dropWhileL' p ('reverse' xs))@.
+dropWhileR :: (a -> Bool) -> StrictSeq a -> StrictSeq a
+dropWhileR p (StrictSeq xs) = StrictSeq (Seq.dropWhileR p xs)
+
+-- | \( O(i) \) where \( i \) is the prefix length.  'spanl', applied to
+-- a predicate @p@ and a sequence @xs@, returns a pair whose first
+-- element is the longest prefix (possibly empty) of @xs@ of elements that
+-- satisfy @p@ and the second element is the remainder of the sequence.
+spanl :: (a -> Bool) -> StrictSeq a -> (StrictSeq a, StrictSeq a)
+spanl p (StrictSeq xs) = toStrictSeqTuple (Seq.spanl p xs)
+
+{-------------------------------------------------------------------------------
+  Indexing
+-------------------------------------------------------------------------------}
+
+-- | \( O(\log(\min(i,n-i))) \). The first @i@ elements of a sequence.
+-- If @i@ is negative, @'take' i s@ yields the empty sequence.
+-- If the sequence contains fewer than @i@ elements, the whole sequence
+-- is returned.
+take :: Int -> StrictSeq a -> StrictSeq a
+take i (StrictSeq xs) = StrictSeq (Seq.take i xs)
+
+-- | \( O(\log(\min(i,n-i))) \). Elements of a sequence after the first @i@.
+-- If @i@ is negative, @'drop' i s@ yields the whole sequence.
+-- If the sequence contains fewer than @i@ elements, the empty sequence
+-- is returned.
+drop :: Int -> StrictSeq a -> StrictSeq a
+drop i (StrictSeq xs) = StrictSeq (Seq.drop i xs)
+
+-- | \( O(\log(\min(i,n-i))) \). Split a sequence at a given position.
+-- @'splitAt' i s = ('take' i s, 'drop' i s)@.
+splitAt :: Int -> StrictSeq a -> (StrictSeq a, StrictSeq a)
+splitAt i (StrictSeq xs) = toStrictSeqTuple (Seq.splitAt i xs)
+
+{-------------------------------------------------------------------------------
+  Helpers
+-------------------------------------------------------------------------------}
+
+toStrictSeqTuple :: (Seq a, Seq a) -> (StrictSeq a, StrictSeq a)
```

Comment:
Doesn't the tuple introduce some laziness again? :thinking:

Reply:
I don't  think so. The point is not that we don't return anything without thunks in it. That's not the goal of a strict type. The goal is that _if_ you force a strict seq to WHNF, _then_ all its elements will be, no more.

Reply:
Yeah, I don't think this is a problem.